airflow/models/dag.py

```diff
@@ -1573,34 +1680,28 @@ def create_dagrun(self,
 
     @classmethod
     @provide_session
-    def bulk_sync_to_db(cls, dags: Collection["DAG"], sync_time=None, session=None):
+    def bulk_sync_to_db(cls, dags: Collection["DAG"], session=None):
```

Comment:
I have doubts about this method:
1. Is the name of this method still correct? `bulk_sync_to_db_and_precompute_fields`?
2. Can a new code not be extracted into new methods/classes?  `SchedulerPredictor`?

Reply:
A separate class doesn't feel right to me -- calculating the values for these fields is very much a property of the dag, and keeping all the logic in one class feels easier to follow than having some part of the dag model updated in a separate class.

This is also why I put the new code in here -- the job of this method is to ensure that the DagModel row in the database is in sync. Since the two new data columns are on the row, I updated them all in one place. 

Particularly because of the `with_for_update` used here (and already in use in master -- not added by my PR) we'd a have to update those columns in the same transaction.

Reply:
Just extracting it to separate methods and changing name would make it better. I think (i wrote about it in the other comment) some of the methods and classes lost their original meaning and they do more or less than originally stated. And the only reason for using separate methods and naming them is to give the developer a chance to understand what those methods do without having to read through all the code. So At least an indication that the methods does more than originally intended would be nice. 

I think where it is easy (such as renames/extractions which modern IDEs make super-easy and 0-risk operation), we should try to make the names reflect a bit more the reality. I agree introducing a new class and separating out the responsibility elsewhere might not be a good thing to do now.

Reply:
I see this as "synchronize the state of the DB with the dags [from the dagbag]" so to me this name accurately reflect the what this method does.

And because of transactions/locking, and wanting to avoid n+1 queries, the best we could do is pull out a method like below, which would still have to be called from within bulk_sync_to_db:

```python
def _calculate_dagrun_date_fields(self, dag, most_recent_dag_run, active_runs_of_dag):

        self.next_dagrun, self.next_dagrun_create_after = dag.next_dagrun_info(
               most_recent_dag_run
        )

        if dag.max_active_runs and active_runs_of_dag >= dag.max_active_runs:
            # Since this happens every time the dag is parsed it would be quite spammy at info
            log.debug(
                "DAG %s is at (or above) max_active_runs (%d of %d), not creating any more runs",
                dag.dag_id, active_runs_of_dag, dag.max_active_runs
            )
            self.next_dagrun_create_after = None

        log.info("Setting next_dagrun for %s to %s", dag.dag_id, self.next_dagrun)
```

Called like this:

```python
            orm_dag._calculate_dagrun_date_fields(
                most_recent_dag_runs.get(dag.dag_id),
                num_active_runs.get(dag.dag_id, 0)
            )
```

Is that the sort of thing you were thinking of?

(Oh except, we can't make it if it's on DagModel as we are calling from within DAG)

Reply:
I think we should change the name, I agree with Kamil and Jarek that it now does a bit more than just "sync_to_db" which for me is like flushing current state to DB.

Maybe `bulk_sync_to_db_and_set_next_dagrun` ? and we could extract the statements that set next dag run to a separate method (`_set_next_dag_run`) that is called from this method.

Reply:
how about `bulk_update_db`? I generally prefer shorter names

Reply:
Yea that works but would be still good to separate the setting of `next_dagrun` to an internal method too and updating the docstrings to reflect that it sets next dagruns too

Reply:
Agree with @kaxil - > separating it out and adding docstring should do the job

Reply:
Done in 63ae0f5c4 - I settled on `bulk_write_to_db` (update could be taken to be UPDATE, but we will INSERT too)